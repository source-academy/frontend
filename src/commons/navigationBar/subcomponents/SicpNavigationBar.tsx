import { Alignment, Drawer, Navbar, NavbarGroup, Position } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import * as React from 'react';
import { useHistory, useParams } from 'react-router';
import controlButton from 'src/commons/ControlButton';
import { getNext, getPrev } from 'src/features/sicp/TableOfContentsHelper';

import { TableOfContentsButton } from '../../../features/sicp/TableOfContentsButton';
import SicpToc from '../../../pages/sicp/subcomponents/SicpToc';

/**
 * Secondary navbar for SICP, should only be displayed when on pages related to interactive /SICP.
 */
const SicpNavigationBar: React.FC = () => {
  const [isTocOpen, setIsTocOpen] = React.useState(false);
  const { section } = useParams<{ section: string }>();
  const history = useHistory();

  const prev = getPrev(section);
  const next = getNext(section);

  const handleCloseToc = () => setIsTocOpen(false);
  const handleOpenToc = () => setIsTocOpen(true);
  const handleNavigation = (sect: string) => {
    history.push('/sicpjs/' + sect);
  };

  // Button to open table of contents
  const tocButton = <TableOfContentsButton key="toc" handleOpenToc={handleOpenToc} />;

  // Previous button only displayed when next page is valid.
  const prevButton = prev && (
    <div key="prev">
      {controlButton('Previous', IconNames.ARROW_LEFT, () => handleNavigation(prev))}
    </div>
  );

  // Next button only displayed when next page is valid.
  const nextButton = next && (
    <div key="next">
      {controlButton('Next', IconNames.ARROW_RIGHT, () => handleNavigation(next), {
        iconOnRight: true
      })}
    </div>
  );

  const drawerProps = {
    onClose: handleCloseToc,
    autoFocus: true,
    canEscapeKeyClose: true,
    canOutsideClickClose: true,
    enforceFocus: true,
    hasBackdrop: true,
    isOpen: isTocOpen,
    position: Position.LEFT,
    usePortal: false
  };

  return (
    <>
      <Navbar className="SicpNavigationBar secondary-navbar">
        <NavbarGroup align={Alignment.LEFT}>{[tocButton]}</NavbarGroup>
        <NavbarGroup align={Alignment.RIGHT}>{[prevButton, nextButton]}</NavbarGroup>
      </Navbar>
      <Drawer {...drawerProps} className="sicp-toc-drawer">
        <SicpToc handleCloseToc={handleCloseToc} />
      </Drawer>
    </>
  );
};

export default SicpNavigationBar;
