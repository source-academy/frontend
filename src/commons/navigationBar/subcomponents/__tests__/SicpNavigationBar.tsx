import { shallow } from 'enzyme';
import ReactRouter from 'react-router';

import SicpNavigationBar from '../SicpNavigationBar';

test('Navbar renders correctly', () => {
  jest.spyOn(ReactRouter, 'useParams').mockReturnValue({ section: 'index' });

  const navbar = <SicpNavigationBar />;
  const tree = shallow(navbar);
  expect(tree.debug()).toMatchSnapshot();
});
