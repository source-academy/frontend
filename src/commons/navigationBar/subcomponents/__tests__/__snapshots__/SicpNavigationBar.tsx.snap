// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders correctly 1`] = `
"<Fragment>
  <Blueprint3.Navbar className=\\"SicpNavigationBar secondary-navbar\\">
    <Blueprint3.NavbarGroup align=\\"left\\">
      <TableOfContentsButton handleOpenToc={[Function: handleOpenToc]} />
    </Blueprint3.NavbarGroup>
    <Blueprint3.NavbarGroup align=\\"right\\">
      <div>
        <Blueprint3.Button disabled={false} fill={false} intent=\\"none\\" minimal={true} className=\\"\\" type={[undefined]} rightIcon={{...}} onClick={[Function (anonymous)]}>
          Next
        </Blueprint3.Button>
      </div>
    </Blueprint3.NavbarGroup>
  </Blueprint3.Navbar>
  <Blueprint3.Drawer onClose={[Function: handleCloseToc]} autoFocus={true} canEscapeKeyClose={true} canOutsideClickClose={true} enforceFocus={true} hasBackdrop={true} isOpen={false} position=\\"left\\" usePortal={false} className=\\"sicp-toc-drawer\\" shouldReturnFocusOnClose={true} style={{...}} vertical={false}>
    <SicpToc handleCloseToc={[Function: handleCloseToc]} />
  </Blueprint3.Drawer>
</Fragment>"
`;
